File: app_movil/lib/models/visit.dart
class Visit {
  final String id;
  final String userId;
  final String monumentId;
  final DateTime date;
  final int? duration; // en minutos
  final int? rating; // 1-5
  final String? device;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Visit({
    required this.id,
    required this.userId,
    required this.monumentId,
    required this.date,
    this.duration,
    this.rating,
    this.device,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Visit.fromJson(Map<String, dynamic> json) {
    return Visit(
      id: json['_id'] ?? json['id'] ?? '',
      userId: json['userId'] ?? '',
      monumentId: json['monumentId'] ?? '',
      date: json['date'] != null
          ? DateTime.parse(json['date'] as String)
          : DateTime.now(),
      duration: json['duration'] as int?,
      rating: json['rating'] as int?,
      device: json['device'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : DateTime.now(),
    );
  }


}

File: backend/src/models/Visit.js
import mongoose from 'mongoose';

const VisitSchema = new mongoose.Schema(
  {
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    monumentId: { type: mongoose.Schema.Types.ObjectId, ref: 'Monument', required: true },
    tourId: { type: mongoose.Schema.Types.ObjectId, ref: 'Tour' }, // Opcional: tour al que pertenece esta visita
    date: { type: Date, default: Date.now },
    duration: { type: Number }, // minutos
    rating: { type: Number, min: 1, max: 5 },
    device: { type: String },
    experienceType: { type: String, enum: ['ar', 'model3d'] },
    clientVisitId: { type: String, trim: true },
  },
  { timestamps: true },
);

// Índices para queries frecuentes
VisitSchema.index({ userId: 1 });
VisitSchema.index({ monumentId: 1 });
VisitSchema.index({ experienceType: 1, date: 1 });
VisitSchema.index({ tourId: 1 });
VisitSchema.index({ userId: 1, tourId: 1 }); // Para obtener todas visitas de un user en un tour
VisitSchema.index(
  { userId: 1, clientVisitId: 1 },
  {
    unique: true,
    partialFilterExpression: { clientVisitId: { $type: 'string' } },
  },
);

export default mongoose.model('Visit', VisitSchema);
